Started implementing DER reader

masihyeganeh opened the pull request:
I could finally find some time to play with your code.
Here is a simple, ugly and maybe not correct implementation of DER reader.
It is easy to write a writer too, but first I need to check with you if I'm in right direction.
Please take a look at this and refactor it to fit better in your project.
I will then complete it.

I have a test that passes with this one, but some values are not right. Maybe it is related to SETs or order of fields.
Unfortunately I can't commit the test, but It would be great if you can write some tests for it.

I needed the SET feature, so I patched them locally, so it may not work without them.

Please help me implement it right.

Thanks

masihyeganeh commented:
Also I did something weird to handle OPTIONAL. please let me know if it is OK or not

kellerkindt commented:
I am currently in  the process of getting SET and SET OF support going - which is now possible since the `TagResolver` in cfb21015ccb255e7032b2dc1eca5f845e88ffe1b has landed.
After #19 has been merged, I will look into this in more detail.
I`ll leave one or two comments on the PR, nonetheless.

kellerkindt commented on src/io/der/octet_aligned.rs:
```diff
@@ -1,7 +1,161 @@
+
+    fn read_identifier(&mut self) -> Result<(Class, PC, u8), Error> {
+
+        let bit8 = self.read_bit()?;
+        let bit7 = self.read_bit()?;
+        let class = Class::from(((bit8 as u8) << 1) + (bit7 as u8));
+
+        let bit6 = self.read_bit()?;
+        let pc = PC::from(bit6);
+
+        let mut tag_bits = vec![0u8; 1];
+        self.read_bits_with_len(&mut tag_bits[..], 5)?;
+        let tag_number = (tag_bits[0] as u8) >> 3;
```
As far as I can tell, DER/BER/CER are designed to read whole octets/bytes and then extract the information from these octets. I imagine something like (totally not tested!):

```rust
let octet = self.read_octet()?;
let bit8 = octet & 0x80 != 0; // required?
let bit7 = octet & 0b40 != 0; // required?
let class = Class:from((octet & 0xC0) >> 2);
```

This would allow us to drop the requirement on `OctetBuffer` for bit-oriented functions and only require simple byte-oriented functions.

masihyeganeh replied:
You are right. I just tried to mimic your code for UPER. Bitwise operation is much better. I'll fix it

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,7 +155,34 @@ impl Reader for DerReader {
         f: F,
     ) -> Result<S, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
```
I don't understand why we are reading here in a `write_enumerated` function.
Since cfb21015ccb255e7032b2dc1eca5f845e88ffe1b `C::Tag` provides the `Tag` of the ENUMERATED instance.

masihyeganeh replied:
It's not actually in `write_enumerated`, Github collapsed some lines here

kellerkindt replied:
Sorry, my bad. You are right

kellerkindt reviewed (commented):
Sorry, but just a hastily review.
Please also add unit tests for encoding/decoding messages.
You may add these as additional encoding tests to already existing uper test structs.

https://asn1.io/asn1playground/ Is a good place to verify the binary output.

masihyeganeh commented on src/io/der/octet_aligned.rs:
```diff
@@ -1,7 +1,161 @@
+impl From<bool> for PC {
+    fn from(v: bool) -> Self {
+        match v {
+            x if x == false => Self::Primitive,
+            x if x == true => Self::Constructed,
+            _ => Self::Constructed
```
My bad, sorry

masihyeganeh commented on src/io/der/octet_aligned.rs:
```diff
@@ -1,7 +1,161 @@
+use crate::io::per::unaligned::BitRead;
 
-impl DistinguishedRead for OctetBuffer {}
+impl BitRead for OctetBuffer {
```
This was another mistake by me because I was trying to mimic your code.
I needed a non-perfect implementation to start correction with you.
I'll change it now

masihyeganeh commented:
Ok, It's now a little better than before.
I just added two tests based on default example of [asn1 playground of ans1.io](https://asn1.io/asn1playground/).
It turn out that there is a lot of problems with tags and big numbers.

I also need to have access to details of field I am reading to be able to validate it. I think we need more data in `C`s.
I hope you can finalize SETs, Tags or anything that you are working on, then we implement this better.

for now, the `simple_der.rs` test passes, but `automatic_tags_der.rs` not.

masihyeganeh commented on src/io/der/octet_aligned.rs:
```diff
@@ -2,6 +2,126 @@ use crate::io::buf::OctetBuffer;
 
-impl DistinguishedRead for OctetBuffer {}
+#[derive(Debug)]
+pub enum Class {
```
Just saw there is an implementation for TAG, Type and Size in `asn1rs-model/src/model/mod.rs`.
I think it's better that I use them, then you provide them in `C` for each parser function, so I can validate them

masihyeganeh replied:
My bad. You just implemented it in [cfb2101](https://github.com/kellerkindt/asn1rs/commit/cfb21015ccb255e7032b2dc1eca5f845e88ffe1b).
The code is a little fragmented for now, I think I need to wait for you to merge all of them, then we continue on top of that

kellerkindt replied:
I just rebased the `experimental-set-setof` branch to include the changes of cfb2101.
I`ll probably merge it into master tomorrow (~20-ish hours from now).

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        eprintln!("[sequence of] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 16 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if !matches!(tag, Tag::Universal(n) if n != 16) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_SEQUENCE != tag {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        let index = tag as u64;
+
+        if let Class::ContextSpecific = &class {
+        } else {
+            return Err(Error::InvalidType)
+        }
```
(With `asn1rs::model::Type`) this can be collapsed to
```rust
if !matches!(tag, Tag::ContextSpecific(_)) {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 2) = (&pc, &tag) {} else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if !matches!(tag, Tag::Universal(2)) || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_INTEGER != tag || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        eprintln!("[utf8string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 12 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 12) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_UTF8_STRING != tag {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        eprintln!("[ia5string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 22 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 22) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_IA5_STRING != tag {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        eprintln!("[octet string] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let 4 = &tag {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 4) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_OCTET_STRING != tag {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on src/syn/io/der.rs:
```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+
+        eprintln!("[boolean] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 1, Length::Definite(1)) = (&pc, &tag, &length) {
+            } else {
+                return Err(Error::InvalidType)
+            }
+        }
```
With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 1) || !matches!(len, LengthDefinitive(1)) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_BOOLEAN != tag || !matches!(length, Length::Definitive(1)) {
    return Err(Error::InvalidType);
}
```

kellerkindt commented on tests/automatic_tags_der.rs:
```diff
@@ -0,0 +1,36 @@
+asn_to_rust!(
+    r"World-Schema DEFINITIONS AUTOMATIC TAGS ::=
+BEGIN
+  Rocket ::= SEQUENCE
+  {
+     range     INTEGER, -- huge (see a special directive above)
+     name      UTF8String (SIZE(1..16)),
+     message   UTF8String OPTIONAL,
+     fuel      ENUMERATED {solid, liquid, gas},
+     speed     CHOICE
+     {
+        mph    INTEGER,
+        kmph   INTEGER
+     }  OPTIONAL,
+     payload   SEQUENCE OF UTF8String
+  }
```
As much as I like this example, for legal reasons I cannot accept their example in this repo

kellerkindt commented on tests/simple_der.rs:
```diff
@@ -0,0 +1,36 @@
+asn_to_rust!(
+    r"World-Schema DEFINITIONS ::=
+BEGIN
+  Rocket ::= SEQUENCE
+  {
+     range     INTEGER, -- huge (see a special directive above)
+     name      UTF8String (SIZE(1..16)),
+     message   UTF8String OPTIONAL ,
+     fuel      ENUMERATED {solid, liquid, gas},
+     speed     CHOICE
+     {
+        mph    [0] INTEGER,
+        kmph   [1] INTEGER
+     }  OPTIONAL,
+     payload   SEQUENCE OF UTF8String
+  }
```
As much as I like this example, for legal reasons I cannot accept their example in this repo

kellerkindt reviewed (changes requested):
Added a few more notes. Do you mind adding a few more tests? For example, if you look in [` tests/basic_integer.rs: 49` at `fn test_uper_small()`](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_integer.rs#L49), it would be really nice to have equivalents for DER (like `fn test_der_small()`). The file [tests/basic_ia5string.rs](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs) is also a good place for inspiration.

Github Actions should also check this PR now automatically.

masihyeganeh commented:
I think you need to rebase this branch on master, then we can do your changes

kellerkindt commented:
The branch has already been merged on master :)

masihyeganeh commented:
I mean the `der-support` branch. it needs to be rebased on master. Am I wrong?

kellerkindt commented:
Oh, yes, you are right! Somehow I thought you this PR was for the `master` branch.

masihyeganeh commented:
> Added a few more notes. Generally, I think you are on the right track 👍
> 
> Do you mind adding a few more tests? For example, if you look in [` tests/basic_integer.rs: 49` at `fn test_uper_small()`](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_integer.rs#L49), it would be really nice to have equivalents for DER (like `fn test_der_small()`). The file [tests/basic_ia5string.rs](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs) is also a good place for inspiration.
> 
> Github Actions should also check this PR now automatically.
> 
> PS: Please run rustfmt or cargo fmt, the CI won't accept it otherwise.

Ok, I rebased my code, changed things you said and made sure that tests passes.
It is far from complete, so please let the tests be these for now. I need to implement `write` to be able to test better.

For the next step, I want to do this:
Reader should should compare current tag with expected one and return `InvalidTag` error if they are not identical.
It may not be needed but I used this trick to implement optional.
For example if schema requires next tag to be an optional utf8string but instead it reads a number, that should return `InvalidTag` error, and the optional parser should receive that, reset read position and return None.
Please tell me if it is OK by you and if there is a better way

kellerkindt commented on src/io/per/err.rs:
```diff
@@ -13,6 +13,7 @@ pub enum Error {
     SizeNotInRange(u64, u64, u64),
     OptFlagsExhausted,
     EndOfStream,
+    InvalidType,
```
I'like to see more debug info here. A `InvalidType(String)` in which the `String` contains something like `"Got unexpected tag {:?}"` or so.

masihyeganeh replied:
Yup. Now that we have the `C::TAG`, it is better to change it to that

kellerkindt commented on tests/simple_der.rs:
```diff
@@ -0,0 +1,39 @@
+
+#[test]
+fn simple_der() {
+    let der_content = b"0'\x02\x0209\x0c\tSomething\n\x01\x00\xa1\x05\x02\x03\x00\xd410\x0c\x0c\x04abcd\x0c\x04efgh".to_vec();
```
Thanks for changing the example. In the long term, I prefer this to have the same notation as in the [remaining crate](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs#L36)

kellerkindt commented on tests/automatic_tags_der.rs:
```diff
@@ -0,0 +1,42 @@
+
+#[test]
+fn automatic_tags_der() {
+    let der_content = b"0'\x80\x0209\x81\tSomething\x83\x01\x00\xa4\x05\x81\x03\x00\xd41\xa5\x0c\x0c\x04abcd\x0c\x04efgh".to_vec();
```
Thanks for changing the example. In the long term, I prefer this to have the same notation as in the [remaining crate](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs#L36)

masihyeganeh replied:
Sure. After implementing `Write`, I'll write separate tests for each data type like existing tests

kellerkindt reviewed (commented):
Nice, Thanks! So what's up next? I guess merging this into the `der-support` branch so you can focus on the writer impl which will be a follow-up PR?

Just as a side question: Is it valid DER to change the order of the elements while serializing? If so, the reader would need to be able to handle reordering the out-of-order elements in the long term? (If so, this could be a separate issue or a follow-up PR)

masihyeganeh commented on src/io/der/octet_aligned.rs:
```diff
@@ -1,7 +1,114 @@
+            1 => Tag::Application(tag_number),
+            2 => Tag::ContextSpecific(tag_number),
+            3 => Tag::Private(tag_number),
+            _ => return Err(Error::UnsupportedOperation("Unsupported tag".to_string())),
```
You are right. I'm trying hard to make it the same as your codes and it won't let me think about these simple things 😆 
I'll change it

masihyeganeh commented on tests/automatic_tags_der.rs:
```diff
@@ -0,0 +1,42 @@
+pub use asn1rs::prelude::*;
+
+asn_to_rust!(
+    r"Data-Structures DEFINITIONS AUTOMATIC TAGS ::=
```
Do you have a plan to implement this `AUTOMATIC TAGS`?

kellerkindt replied:
To be honest, I didn't look into the Tag stuff too much, since it was never relevant for uPER. I am open to it being implemented.

masihyeganeh commented:
> Nice, Thanks! So what's up next? I guess merging this into the `der-support` branch so you can focus on the writer impl which will be a follow-up PR?
> 
> Just as a side question: Is it valid DER to change the order of the elements while serializing? If so, the reader would need to be able to handle reordering the out-of-order elements in the long term? (If so, this could be a separate issue or a follow-up PR)

Yes, I think it is OK for this PR to be merged (after implementing changes you wrote above) and continue on `Writer` in another PR.

I don't think it is valid in any of BER/CER/DER/XER/UPER/... to change order of elements (I may be wrong). I think it's only valid in SETs (I may be wrong about this too)

masihyeganeh commented:
Ok, I think you can merge it now if you are OK with it.
I'll continue in another PR.
I think there is a problem with TAG implementation.
I'll work on that.

kellerkindt reviewed (approved):
LFTM
